kani-compiler/src/kani_middle/transform/loop_contracts.rs

```diff
@@ -0,0 +1,266 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! This module contains code related to the MIR-to-MIR pass to enable loop contracts.
+//!
+
+use crate::kani_middle::codegen_units::CodegenUnit;
+use crate::kani_middle::find_fn_def;
+use crate::kani_middle::transform::body::{MutableBody, SourceInstruction};
+use crate::kani_middle::transform::{TransformPass, TransformationType};
+use crate::kani_middle::KaniAttributes;
+use crate::kani_queries::QueryDb;
+use crate::stable_mir::CrateDef;
+use rustc_middle::ty::TyCtxt;
+use rustc_span::Symbol;
+use stable_mir::mir::mono::Instance;
+use stable_mir::mir::{BasicBlockIdx, Body, Operand, Terminator, TerminatorKind};
+use stable_mir::ty::{FnDef, RigidTy};
+use stable_mir::DefId;
+use std::collections::VecDeque;
+use std::collections::{HashMap, HashSet};
+use std::fmt::Debug;
+use tracing::trace;
+
+/// This pass will perform the following operations:
+/// 1. Replace the body of `kani_register_loop_contract` by `kani::internal::run_contract_fn`
+///    to invoke the closure.
+///
+/// 2. Replace the dummy call to the register function with the actual call, i.e., transform
+///
+/// ```ignore
+/// let kani_loop_invariant = || -> bool {inv};
+/// kani_register_loop_contract(kani_loop_invariant)
+/// while guard {
+///     loop_body;
+///     kani_register_loop_contract(||->bool{true});
+/// }
+/// ```
+///
+///    to
+///
+/// ```ignore
+/// let kani_loop_invariant = || -> bool {inv};
+/// while guard {
+///     loop_body;
+///     kani_register_loop_contract(kani_loop_invariant);
+/// }
+///
+/// ```
+///
+/// 3. Move the call to the register function to the loop latch terminator. This is required
+///    as in MIR, there could be some `StorageDead` statements between register calls and
+///    loop latches.
+#[derive(Debug, Default)]
+pub struct FunctionWithLoopContractPass {
+    /// Cache KaniRunContract function used to implement contracts.
+    run_contract_fn: Option<FnDef>,
+    /// Function and Arguments of register functions.
+    registered_args: HashMap<DefId, (Operand, Vec<Operand>)>,
+    /// The terminator we are moving to the loop latch.
+    loop_terminator: Option<Terminator>,
+}
+
+impl TransformPass for FunctionWithLoopContractPass {
+    fn transformation_type() -> TransformationType
+    where
+        Self: Sized,
+    {
+        TransformationType::Stubbing
+    }
+
+    fn is_enabled(&self, _query_db: &QueryDb) -> bool
+    where
+        Self: Sized,
+    {
+        true
```

Comment:
What happens if the loop invariant check hasn't been enabled?

Reply:
It should be true only when loop contracts is enable. Updated.